### DISCUSSION TITLE
Recommended basis for strong, stable etags?

### DISCUSSION THREAD

**Discussion by trevhunter** [ravendb/ravendb#14089] [Q&A]
In RavenDB, we've generally used [Change Vectors](https://ravendb.net/docs/article-page/5.3/Csharp/server/clustering/replication/change-vector) as [HTTP Etags](https://en.wikipedia.org/wiki/HTTP_ETag) in our applications. 

This has let us effectively implement [Conditional GET's](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match) and even [Conditional updates](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Match) all the way to the database level in an efficient way. Even RavenDB itself [uses the Change Vector like this](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Match).

But we've kinda been using them naively, because there's some important things we've overlooked:

* Change vectors are a composition of the values from multiple nodes. I guess changing the cluster topology may result in all Change Vectors being updated, even if the data hasn't actually changed. I haven't confirmed this behavior though. 
* Change vectors are not maintained in certain circumstances (e.g. restore from non-snapshot backup, Export/Import).

In other words, certain disaster recovery or maintenance operations might cause all previously issued etags to be invalid. And this could potentially result in a spike of load as end clients and various caches all invalidate at the same time, undoing all the efficiencies gained by conditional gets. 

What's the recommended best practice for end-to-end ETags on documents, and how best to map from change vector to etag? Is there a solution that keeps etags stable without resorting to an app-generated value stored on the doc itself?

**Comment by ayende** [comment]
* A change vector will _only_ change when you are changing the document, not when you add a new node to the cluster. That will change the change vector of _modified documents_, of course, as is expected.
* Change vectors with backup/restore or import/export aren't preserved (effectively, we update them over empty database, so new change vector). The reasoning here is that you have a _new_ database, not the same as the old one. 

The scenario for conditional updates & conditional gets don't really apply here, since you just reset the entire persistent ecosystem.

From your perspective, the issue is long term client "memories", right? So a client may come back a week later and ask for updates, and it is important to avoid the change if you need to restore from scratch.

From a DR perspective, there are a few things to note:

* Use snapshot for backups, which preserve change vectors (and are cheaper to restore).
* Use a separate cluster for DR (external replication), since replication preserve the change vectors.

Note that the last-modified metadata property is also updated on restore from backup or import.

**Reply by trevhunter** [reply]
Thanks @ayende 

I'll preface this: The originating issue (Etag stability) is a surprise, but not a critical concern. I'm pretty confident that our apps and RavenDB could handle the load of millions of clients doing a full sync and that no logic errors would result. But I would love to remove the "surprise" aspects :-)

A few thoughts on this:

> The reasoning here is that you have a new database, not the same as the old one.

It was a bit of a surprise that "backups" don't preserve all data (including system owned metadata such as change vectors and last modified date). I can understand that an *Export/Import* is effectively operating on a new DB, but semantically, a *Backup/Restore* (of any type - snapshot, full, incremental) feels like it should restore *all state* as it was. 

Perhaps a bit more detail [in the backup docs](https://ravendb.net/docs/article-page/5.3/csharp/server/ongoing-tasks/backup-overview) explaining this nuance of "logical backups" is warranted. It's implied in stating that *"A logical backup is a compressed JSON dump of database contents, including documents and additional data."* but clarifying which "additional data" is preserved and which isn't may be useful, and reflecting it in the [change vector docs](https://ravendb.net/docs/article-page/5.3/Csharp/server/clustering/replication/change-vector) too may be helpful. 

> Use snapshot for backups, which preserve change vectors (and are cheaper to restore).

I understand why a snapshot is needed, given the above point about other types of backup not preserving change vector, but the overhead (license type, size, time to take a backup, IO load on the node taking the backup) means it's not something we'd do as often as the incremental backups (although time to restore is a benefit).

In terms of our DR strategy, we have a fairly common 3 tier setup:
1. *High Availability*: 3+ nodes in a production cluster (with the hosts in different fault zones, but same datacenter). 
2. *Disaster Recovery*: External replication to a different cluster in a different data center. 
3. *Big Disaster Recovery*: Offsite backups - currently full backup infrequently, incremental backup frequently. 

If we hit scenario 3 with needing the offsite backups, we have big problems in general and every little surprise matters. But we also sometimes use those backups to restore to a test cluster for debug & analysis purposes so differences in data and metadata may obscure some of those things. 

Ideally (on order of preference)

1. All backup types would restore all state of documents - data and metadata. 
2. Snapshot backups come with a few more options: (a) Available on professional tier license (b) Optionally exclude index data.
3. More explicitly detail the "Logical Backup" limits & considerations in the docs and the UI as outlined above.

**Reply by ayende** [reply]
FWIW - backup / restore is just import/export under the covers. We _have_ the information about the previous change vector / modified date.
At this time, I cannot recall why we are doing the change there. I don't think that we care too much about it, but I'm not certain.

@arek / @ppekrol / @karmeli87 thoughts?

**Reply by karmeli87** [reply]
Change vector is not only used to handle versioning of a certain document, but it also used in a more global context of a partial order of the _entire_ database.
e.g. document A:100 can't appear before a document with A: 42 (different documents).

Retaining the change vector while importing might mess this up. If you import a file with A:100 first and then a different file with A;42, tasks like ETL, Replication, Subscription etc might skip the A;42, because they first will process A:100 and will consider the A:42 as already processed because 100 > 42.

Hope it make sense :-)

**Reply by trevhunter** [reply]
Hi @karmeli87 - this makes sense - specifically in the context of an *import*. But it also doesn't seem like an unreasonable expectation to limit *backups* to being (a) One time import (not multiple imports) (b) Ordered.

**Reply by ayende** [reply]
In that context, the problem is that we have a problem with the following scenario:

* User restores from backup
* User makes some changes
* User restores from (same) backup

At this point, the change vectors we have are both older and newer than what we have in place. 
That is a recipe for high level of complexity.

**Reply by trevhunter** [reply]
I'm confused... is that a *real* use case? In the 3rd step you outline, I would have deleted the DB used in step 1&2 and don't care about the changes I've made in step 2.

When restoring from backup, I don't expect to restore *on top of an existing database*. A restore from backup is used to create a database that is an copy of the db that was backed up. I'd use an *import* if I wanted to add/merge data to an existing DB. 

In the studio today, I can easily see the "create new db from backup" option, but I there's only an "import data" task for an existing DB, so there already seems to be a guard against restoring a backup on top of something existing.

**Reply by ayende** [reply]
That is a scenario that we worry about, and it is real. 
A user may want to reset the database back to its original state. They do that by restoring on top of already existing database.
What happens at this point is that you have overwritten all the documents in the backup, but any _other_ documents remains the same.

Common scenario for this is when you are reproducing an issue, when you go back to the same state each time.

**Reply by trevhunter** [reply]
> A user may want to reset the database back to its original state. They do that by restoring on top of already existing database.

Why do this *on top of an already existing db* vs just restoring it fresh? I don't see how restoring it on top of an existing DB gets things back to its *original* state. If there are changes after the point that the backup captured that need to remain, then that's not its original state. 

I'm missing something in this use case, but even so, shouldn't it still be possible to have:
(a) Ordered data in backup files.
(b) When restoring to a new database, maintain change vectors and etags (preserving all state)
(c) If restoring on top of an existing db, provide a warning about how this won't maintain certain things like change vectors and won't overwrite changes in the target db since the backup.

**Reply by ayende** [reply]
What you consider valid and the support scenarios that we have to resolve aren't necessary a match, I'm afraid.
I agree that this is a bad scenario, however, that doesn't mean that we don't have to field support calls on that.

Recently we had a scenario where a user restore a database around 500 times (as part of a test scenario) using the same database as the source. There was a problem with merging the existing and new database record and the database record hit 500MB (!).
That is a bug in our code, but it is a scenario users are running into.

**Reply by mcrio** [reply]
So they may make some changes to the document structure like adding new fields, and when restoring the backup on top of an existing DB it would just merge the known fields? Asking as I didn't understand the scenario as well.

**Reply by ayende** [reply]
Create new documents, usually.
Modify existing docs / delete existing docs.

Then you have a restore on top of that to have the original docs back, but keep the new docs as well.

**Reply by trevhunter** [reply]
Given that it's a real use case (my mistake), perhaps then it's just a case of feature overloading. To me, this would make more sense and be in line with what I'd expect semantically:

1. *Backup/Restore* always restores all original state. It always starts from a **new & blank** database. All the state is restored as it was according to the backup. Indeed, in the studio, you only have the option to have "New Database from backup (restore)" option, but no (explicit) "Restore on top of current database" option[^1]. 

2. *Import* will take some file (either a backup file or an export) and merge that state with an **existing** database. Given that exports and non-snapshot backups are the same file format, I'm guessing it wouldn't be hard to allow the import feature to use a backup file as its source. 

Explicitly separating the behaviors of restoring original state (backup/restore) and merging state (import) feels like it simplifies things overall, keeps both use cases, could be documented separately, and is maybe more in line with general expectations of the terms. 

[^1]: Yes, I'm aware of the "New database, but if there are files on disk with the same name use them" idea when creating a new database, but I've always wondered why that vs "New, blank db" aren't explicitly separate features.

**Reply by ayende** [reply]
That is an option, but I'm not sure that this sort of change (which is pretty drastic) is advisable.

To answer you other question, because it relates, *not* blocking those sort of this (new db on existing files) enables you to do some things that you couldn't do otherwise. 

For example, let's say you want to do "online" compression of documents. Right now, after enabling compression, you have to run compaction (which can take a while). So you can do:
* create new db `my-db-new`
* setup compression
* external replication to new db from old
* let it run
* when done, delete the old db
* create it again, pointing to the new path

You just reduced the time it was down significantly.

The idea is to _not_ block behavior unless is is known to be problematic, because otherwise you need to create specialized behavior for each and every scenario.

In the same sense, adding specialized behavior for restore-on-new-db vs restore-on-existing-db starts carving out a lot of complex exceptions.

**Reply by ayende** [reply]
One last thing, I was reminded that we have the following format internally to the backup file:

```
"Docs":[{...}],
"Revisions":[{...}],
"TimeSeries":[{...}],
```

That means that the write _order_ isn't sorted between Docs & Revisions on restore. 
Preserving the change vector in this case is not going to be workable. 

I think that the recommendation in this scenario is to use snapshots if change vector stability on restore matters.

**Reply by garayx** [reply]
> Perhaps a bit more detail [in the backup docs](https://ravendb.net/docs/article-page/5.3/csharp/server/ongoing-tasks/backup-overview) explaining this nuance of "logical backups" is warranted. It's implied in stating that _"A logical backup is a compressed JSON dump of database contents, including documents and additional data."_ but clarifying which "additional data" is preserved and which isn't may be useful, and reflecting it in the [change vector docs](https://ravendb.net/docs/article-page/5.3/Csharp/server/clustering/replication/change-vector) too may be helpful.

We will add this info, thanks for suggestion, 
[https://issues.hibernatingrhinos.com/issue/RDoc-2175](https://issues.hibernatingrhinos.com/issue/RDoc-2175)

**Reply by trevhunter** [reply]
Thanks @garayx and @ayende .

> I think that the recommendation in this scenario is to use snapshots if change vector stability on restore matters.

This isn't really feasible: (a) Snapshots on a very large DB are problematic to run frequently (b) They require an enterprise license.

It seems that the only lightweight way to preserve it is to have an application-defined `etag` that isn't based on `Change Vector` but the downside of that is we could no longer do the conditional gets all the way to the DB. Oh, well... tradeoffs I guess.

**Comment by georgiosd** [comment]
@trevhunter I've been thinking about this for a while now, since I observed express+mongo doing this somehow automagically. How did you go about implementing it? Presumably in some version of ASP.NET?

**Reply by trevhunter** [reply]
@georgiosd Apologies, I'm not sure exactly what you mean... Are you wondering how we did the original HTTP etag based on change vectors or how we would aim to solve it without change vectors?

**Reply by georgiosd** [reply]
The first. And if I understand correctly you did for "diff" reasons, but whether it gave you a noticeable boost in performance (if you measured it).

**Reply by trevhunter** [reply]
@georgiosd No automagic or even magic :-)

We have a HTTP API for data that serves millions of devices. It's similar to that [described in this whitepaper.](https://ravendb.net/whitepapers/couchbase-vs-ravendb-performance-at-rakuten-kobo)

We use eTags for two things:
1. Performance and load shedding: We use conditional gets to avoid unnecessarily sending the payload to the client and we use eTags to avoid making further calls altogether (if a root entity hasn't changed, the client doesn't bother making calls to refresh related things).
2. Optimistic concurrency. If a client is going to do a `PATCH` or `DELETE`, it sends the etag. The operation fails if the client etag is different. 

Our application API is built on ASP.Net Core 6. 

* When retrieving an entity using a `GET`, the client sends its current eTag (if it has one) as an `If-None-Match` header. We [pass this to RavenDB for a conditional get](https://ravendb.net/docs/article-page/5.3/csharp/client-api/session/loading-entities#conditionalload). If the document doesn't get loaded, we send back a `HTTP 304 Not Modified`. When we do have changes to send, we return the current change vector as an eTag response header.
* When updating an entity with `PATCH` or `DELETE` the client sends its current eTag as an `If-Match` header. If the eTag doesn't match the change vector, we send back `HTTP 412 Precondition Failed` to the client. 

All of the above is a pretty simple mapping to the HTTP Request and Response headers in the controller. 

As for a "noticeable performance boost":
* The conditional GET itself is a very minor operation. The documents are very small, so not returning the body doesn't really affect latency much (if at all). Over billions of requests though where > 99% are `HTTP 304`, the small bandwidth savings add up. On the client side, when you have a limited power device (as in our eReaders, but also common in IOT), avoiding using the CPU to unnesseciarily parse the JSON and update non-volatile storage is a boon for both performance and battery life. 
* By far the largest saving on the server side is the load shedding - the sync mechanism that the clients use can avoid multiple subsequent calls if the data hasn't changed. This is a huge boon for overall scalability. 

Hope this helps.

**Reply by georgiosd** [reply]
@trevhunter Never thanked you for the elaborate response Trevor! Thank you, a lot to chew on here :)
